sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPredicates.scala

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.spark
+
+import com.azure.cosmos.implementation.Strings
+import org.apache.spark.SparkEnv
+
+private object CosmosPredicates {
+  private[this] val ParameterName = "parameterName"
+
+  /**
+   * Executor id for the driver.  In earlier versions of Spark, this was `<driver>`, but this was
+   * changed to `driver` because the angle brackets caused escaping issues in URLs and XML (see
+   * SPARK-6716 for more details).
+   */
+  private[this] val DRIVER_IDENTIFIER = "driver"
+
+  private[this] def argumentMustNotBeNullOrEmptyMessage(parameterName: String): String =
+    s"Argument '$parameterName' must not be null or empty."
+
+  private[spark] def requireNotNull[T](candidate: T, parameterName: String): T = {
+    requireNotNullOrEmpty(parameterName, ParameterName)
+    require(candidate != null, s"Argument '$parameterName' must not be null.")
+    candidate
+  }
+
+  private[spark] def requireNotNullOrEmpty(candidate: String, parameterName: String): String = {
+    require(
+      !Strings.isNullOrWhiteSpace(parameterName),
+      argumentMustNotBeNullOrEmptyMessage(ParameterName))
+    require(!Strings.isNullOrWhiteSpace(candidate), argumentMustNotBeNullOrEmptyMessage(parameterName))
+    candidate
+  }
+
+  private[spark] def assertNotNullOrEmpty(candidate: String, parameterName: String): String = {
+    assert(
+      !Strings.isNullOrWhiteSpace(parameterName),
+      argumentMustNotBeNullOrEmptyMessage(ParameterName))
+    assert(!Strings.isNullOrWhiteSpace(candidate), argumentMustNotBeNullOrEmptyMessage(parameterName))
+    candidate
+  }
+
+  private[spark] def assertNotNull[T](candidate: T, parameterName: String): T = {
+    assertNotNullOrEmpty(parameterName, ParameterName)
+    assert(candidate != null, s"Argument '$parameterName' must not be null.")
+    candidate
+  }
+
+  private[spark] def requireNotNullOrEmpty[T](candidate: Array[T], parameterName: String): Array[T] = {
+    require(!Strings.isNullOrWhiteSpace(parameterName), argumentMustNotBeNullOrEmptyMessage(ParameterName))
+    require(candidate != null && !candidate.isEmpty, argumentMustNotBeNullOrEmptyMessage(parameterName))
+    candidate
+  }
+
+  private[spark] def assertNotNullOrEmpty[T](candidate: Array[T], parameterName: String): Array[T] = {
+    assert(!Strings.isNullOrWhiteSpace(parameterName), argumentMustNotBeNullOrEmptyMessage(ParameterName))
+    assert(candidate != null && !candidate.isEmpty, argumentMustNotBeNullOrEmptyMessage(parameterName))
+    candidate
+  }
+
+  private[spark] def assertOnSparkDriver(): Unit = {
+    // assert that we're only accessing it on the driver.
+    assert(SparkEnv.get.executorId == DRIVER_IDENTIFIER, "This code should only be executed on the Spark driver.")
```

Comment:
https://github.com/apache/spark/blob/415506cc04f383c7006f46c8b9dd0ddf0f762d7d/core/src/main/scala/org/apache/spark/SparkContext.scala#L2769-L2774

Reply:
https://github.com/apache/spark/blob/8a1e172b513ba58763336de83f94e00ceaa69255/core/src/main/scala/org/apache/spark/SparkEnv.scala#L402-L405

Reply:
It is not exposed in public API - but as a environment setting in publicly available key/value config. I don't think this will change without being treated as breaking change in Spark.  So I considered it safe.